8.0 account_invoice_accrual

sbidoul opened the pull request:
This new module lets the user generate accrual move for supplier (or customer) invoices with one click on the draft invoice form.

When goods or services are received, the draft invoice is generated using standard odoo processes. If at the end of the period, the supplier invoice has not been received, the user can click "Accrue" on the invoice to generate an accrual move in the period being closed. Next period when the supplier invoice arrives, the accrual move is automatically reversed when the invoice is valided.

It is based on account_cutoff_accrual_base account mapping framework.

This PR is still WIP while we fix the last issues being discovered during testing.

This PR depends on #8

Commit:
Add account_invoice_accrual module

Commit:
[FIX][account_invoice_accrual] Crash without model=model_name

Commit:
[IMP][account_invoice_accrual] view_load attribute is deprecated on ir.fields

Commit:
[FIX][account_invoice_accrual] Some fixes to make yml tests happy

Commit:
[FIX][account_invoice_accrual] Flake8 and pylint issues

Commit:
[IMP][account_invoice_accrual] Bad display for account_accrual_move_id

Commit:
[IMP][account_invoice_accrual] Avoid to copy accural_move_id

Commit:
[FIX][account_invoice_accrual] Type isn't initialized

Commit:
[IMP][account_invoice_accrual] Use accrual date to set date on analytic line not invoice date that can be not initialized

Commit:
[ADD][account_invoice_accrual] Add README

Commit:
[IMP][account_invoice_accrual] Account_invoice_accrual refactoring :
* Remove accrual information on res_partner
* Use default cutoff journal which is on company
* Use accrual accounts defined on account_cutoff_accraul_base
* Use accounts mapping which is implemented in account_cutoff_base
* Adapt test suite

Commit:
[ADD][account_invoice_accrual] Remove/reversal accrual :
* Remove accrual in case of cancel or unlink invoice if accrual entry is draft
* Remove accrual in case of invoice's validation if accrual entry is draft and if entry genereted by the validation is on the same period as accrual
* Otherwise, accrual entry is reversed

Commit:
[ADD][account_invoice_accrual] Add the possibility to consider taxes on accrual or not

Commit:
[IMP][account_invoice_accrual] set to_be_reversal to true after accrual

Commit:
[IMP][account_invoice_accrual] Display invoice's origin on accrual aml's description

yvaucher commented on account_invoice_accrual/__openerp__.py:
```diff
@@ -0,0 +1,64 @@
+        "account_reversal",  # from account-financial-tools/7.0
+        "account_cutoff_accrual_base",
+    ],
+    "description": """
```
You can remove this from `__openerp__.py`

yvaucher commented on account_invoice_accrual/__openerp__.py:
```diff
@@ -0,0 +1,64 @@
+{
+    "name": "Account invoice accrual",
+    "version": "0.1",
+    "author": "ACSONE SA/NV",
```
Please add OCA in author list

yvaucher commented on account_invoice_accrual/__openerp__.py:
```diff
@@ -0,0 +1,64 @@
+        "test/account_invoice_accrual_remove.yml",
+        "test/account_invoice_accrual_reversal.yml",
+    ],
+    "active": False,
```
active is an alias for auto_install please remove it

yvaucher commented on account_invoice_accrual/wizard/account_move_reverse.py:
```diff
@@ -0,0 +1,88 @@
+            context = {}
+        inv_obj = self.pool.get('account.invoice')
+        invoice_ids = []
+        if context.get('active_model') and \
```
You can use `()` and remove `\` but this is not mandatory

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+
+    def move_line_get_item(self, cr, uid, line, context=None):
+        res = super(account_invoice_line, self)\
+            .move_line_get_item(cr, uid, line, context=None)
```
Context is not propagated

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+                                             context=context)
+
+    def action_cancel(self, cr, uid, ids, context=None):
+        if context is None:
```
No need to check context here as we are not getting anything from it
Check to move under reverse_invoice

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+        return res
+
+    def invoice_validate(self, cr, uid, ids, context=None):
+        if context is None:
```
No need to check context

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+        return res
+
+    def unlink(self, cr, uid, ids, context=None):
+        if context is None:
```
No need to check context

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+
+    def reverse_invoice(self, cr, uid, ids, context=None):
+        # get the list of invoice to reverse
+        ids_to_reverse = []
```
Please check context is dict here

yvaucher commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,327 @@
+
+        # one move line per tax line
+        if accrual_taxes:
+            iml += ait_obj.move_line_get(cr, uid, invoice.id)
```
missing context propagation

yvaucher commented on account_invoice_accrual/i18n/account_invoice_accrual.pot:
```diff
@@ -0,0 +1,204 @@
+
+#. module: account_invoice_accrual
+#: help:account.move.accrue,date:0
+msgid "Enter the date of the accrual account entries. By default, OpenERP proposes the last day of the previous month."
```
s/OpenERP/Odoo

yvaucher commented on account_invoice_accrual/i18n/fr.po:
```diff
@@ -0,0 +1,204 @@
+
+#. module: account_invoice_accrual
+#: help:account.move.accrue,date:0
+msgid "Enter the date of the accrual account entries. By default, OpenERP proposes the last day of the previous month."
```
s/OpenERP/Odoo

yvaucher commented on account_invoice_accrual/i18n/fr.po:
```diff
@@ -0,0 +1,204 @@
+#. module: account_invoice_accrual
+#: help:account.move.accrue,date:0
+msgid "Enter the date of the accrual account entries. By default, OpenERP proposes the last day of the previous month."
+msgstr "Spécifier la date de provision. Par défaut, OpenERP propose le dernier jour du mois précédent."
```
s/OpenERP/Odoo

yvaucher commented on account_invoice_accrual/wizard/account_move_accrue.py:
```diff
@@ -0,0 +1,143 @@
+            'Accrual Date',
+            required=True,
+            help="Enter the date of the accrual account entries. "
+                 "By default, OpenERP proposes the last day of "
```
s/OpenERP/Odoo

gurneyalex commented:
I just restarted the travis build.

Commit:
[FIX][account_invoice_accrual] Flake8 issue

Commit:
[DEL][account_invoice_accrual] Remove description for manifest

Commit:
[ADD][account_invoice_accrual] Add OCA as author

Commit:
[DEL][account_invoice_accrual] Remove active key on the manifest

Commit:
[FIX][account_invoice_accrual] Context isn't propagated

Commit:
[IMP][account_invoice_accrual] No need to check contest here.

Commit:
[FIX][account_invoice_accrual] Check context

Commit:
[FIX][account_invoice_accrual] Missing context propagation

Commit:
[IMP][account_invoice_accrual] Replace OpenERP by Odoo

Commit:
[ADD][account_invoice_accrual] Add new account on YAML test

Commit:
[FIX][account_invoice_accrual] Set default date for accrual wizard on YAML test

Commit:
[FIX][account_invoice_accrual] Use default value for journal on accrual wizard in YAML tests

adrienpeiffer commented:
@yvaucher Thanks for your review !

yvaucher commented on account_invoice_accrual/README.rst:
```diff
@@ -0,0 +1,31 @@
+Account Invoice Accrual
```
Can you add the license badge in README ?

see https://github.com/OCA/maintainer-tools/tree/master/template/module

adrienpeiffer replied:
Of course !

adrienpeiffer replied:
@yvaucher it's done !

yvaucher commented:
@adrienpeiffer Thanks for the changes just a tiny detail and it is :+1:

Commit:
[IMP][account_invoice_accrual] Add licence badge

adrienpeiffer commented:
Travis is green now :smile:

yvaucher commented:
Thanks we need a second approval there

Maybe @sbidoul can do it ? Then we can merge

By the way I think you can remove the [WIP] in title

sbidoul commented:
@yvaucher I'll do a detailed functional test soon.

sbidoul commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,323 @@
+            readonly=True, ondelete='restrict', copy=False,
+            help="Link to the Accrual Journal Items."),
+        'accrual_move_name': fields.char('Accrual Journal Entry', size=64,
+                                         readonly=True, copy=False),
```
What is the purpose of this accrual_move_name field?

sbidoul commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,323 @@
+    def move_line_get_item(self, cr, uid, line, context=None):
+        res = super(account_invoice_line, self)\
+            .move_line_get_item(cr, uid, line, context=context)
+        if context.get('move_accrual', False) and context.get('type', False):
```
, False not necessary

sbidoul commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,323 @@
+            store=False, readonly=True),
+    }
+
+    def reverse_invoice(self, cr, uid, ids, context=None):
```
better name this reverse_accrual?

adrienpeiffer replied:
Yes, I think too

sbidoul commented on account_invoice_accrual/account_invoice.py:
```diff
@@ -0,0 +1,323 @@
+            wiz_obj.action_reverse(cr, uid, [wizard_id], context=wiz_context)
+        if move_ids_to_unlink:
+            self.write(cr, uid, invoice_to_unlink_move,
+                       {'accrual_move_id': False})
```
I'm not sure we should reset accrual_move_id when reversing. Because if the user reverses the journal entry without going through the reversal button on the invoice, it will not be reset. So that would give a slightly different behaviour, potentially confusing.

sbidoul replied:
Ah, sorry, it works because the account.move.reversal wizard is extended too. Good.

sbidoul commented:
I found a scenario with an issue:
- today = 2015-05-05
- create supplier invoice dated 2015-04-15 with period 04/2015
- accrual 2015-04-30 -> ok
- change invoice period to 05/2015
- validate invoice -> the reveral move takes date 2015-06-01 and period 06/2015: NOK

When validating the invoice, the reversal move must have the same period as the invoice move (and I'd say the first day of the period).

Commit:
[FIX][account_invoice_accrual] Ensure reversal period is the same as invoice's move in case of invoice's validation

Commit:
[IMP][account_invoice_accrual] Rename reverse_invoice to reverse_accrual

Commit:
[IMP][account_invoice_accrual] False isn't necessary

adrienpeiffer commented:
Rebased to get travis file changes

Commit:
[ADD][account_invoice_accrual] Use accrual default journal

adrienpeiffer commented:
- [x] Depends on #24

Commit:
[IMP][account_invoice_accrual] Accrual_move_name isn't necessary

sbidoul commented:
The build is red because it depends on #24 but is otherwise ok.

sbidoul committed:
[FIX] account_invoice_accrual: typo in reversal button domain

sbidoul committed:
[IMP] account_invoice_accrual: no need to prevent deletion of accrual moves

sbidoul committed:
[FIX] account_invoice_accrual: accrual_move_name is really not necessary

sbidoul committed:
[FIX] account_invoice_accrual: update copyright headers

sbidoul commented:
I discovered another issue when using hr_timesheet_invoice. Analytic lines linked to accrual moves are proposed for invoicing because account_move_line.invoice (which is a computed field) is False for accrual moves and therefore analytic lines are not linked to the invoice (https://github.com/odoo/odoo/blob/8.0/addons/hr_timesheet_invoice/hr_timesheet_invoice.py#L363).

This module is definitely harder to get 100% right than I expected.

sbidoul committed:
[IMP] account_invoice_accrual: filter closed and special periods

sbidoul committed:
[IMP] account_invoice_accrual: simplify and partially migrate to new api

Main motivation for this change is that the override of account_reversal
is fragile, error prone and may not work in all situations (in particular
when reversal are done with another mean than the wizard). So it's
better to leave the link to accrual_move_id even after reversal: it
does no harm.

sbidoul committed:
[IMP] account_invoice_accrual: restore the Reverse Accrual button

This is just a shortcut to the Reverse wizard on the accrual move.

sbidoul committed:
[IMP] account_invoice_accrual: improve README.rst

Commit:
[ADD][account_invoice_accrual] Get the correct default journal that depends on invoice's type

Commit:
[IMP][account_invoice_accrual] Improve indentation

sbidoul committed:
account_invoice_accrual: normalize version number

sbidoul committed:
[ADD] setup.py

adrienpeiffer commented:
Work fine in production for more than 1 year

lmignon commented:
Work fine in production for more than 1 year

adrienpeiffer commented:
Thanks @yvaucher !

yvaucher commented:
@sbidoul @adrienpeiffer @lmignon This merge make the travis tests fail:
https://travis-ci.org/OCA/account-closing/jobs/162209582

Can you check ? It seems it is incompatible with `account_reversal` module

sbidoul commented:
This is due to a recent change by @alexis-via in account_reversal https://github.com/OCA/account-financial-tools/pull/387 that came after the last travis run of this module PR.

sbidoul commented:
I'm working on it

sbidoul commented:
#38
